### PR TITLE
refactor(billing): store `payment_method_data_billing` for recurring payments

### PR DESCRIPTION
**Pull request by Narayanbhat166**
## Type of Change

- [x] Refactoring

## Description

When a card or any payment method is saved, the billing address associated with it is also stored. This has to be presented back when the payment method is listed. This PR adds support for storing and retrieving the `payment_method_billing` during listing the payment methods associated to a customer.

## How did you test it?

- Create a CIT payment with `payment_method_data.billing`.
```bash
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'api-key: dev_oN54ym4AmrV4om0Cs5Roq2JShISv69npT9BgwdsYABeTkhCmI77V5Gdpeqy7hwWE' \
--data-raw '{
    "amount": 6540,
    "currency": "USD",
    "confirm": true,
    "customer": {
        "id": "cus_abc",
        "email": "example@juspay.in"
    },
    "customer_acceptance": {
        "acceptance_type": "offline",
        "accepted_at": "1963-05-03T04:07:52.723Z",
        "online": {
            "ip_address": "127.0.0.1",
            "user_agent": "amet irure esse"
        }
    },
    "payment_method": "card",
    "payment_method_type": "debit",
    "payment_method_data": {
        "card": {
            "card_number": "4000000000001091",
            "card_exp_month": "12",
            "card_exp_year": "26",
            "card_holder_name": "joseph Doe",
            "card_cvc": "123"
        },
        "billing": {
            "address": {
                "line1": "1467",
                "line2": "Harrison Street",
                "city": "San Fransico",
                "state": "California",
                "zip": "94122",
                "country": "US",
                "first_name": "Narayan",
                "last_name": "Bhat"
            }
        }
    },
    "setup_future_usage": "off_session"
}'
```
- Do list customer payment methods, and ensure that the billing field is populated
```bash
curl --location 'http://localhost:8080/customers/payment_methods?client_secret=pay_Dv3zMLDiCbpiVY3bgwXB_secret_33wrwysfn34oFovDKio8' \
--header 'Accept: application/json' \
--header 'api-key: pk_dev_3c59efb2965a45c980c044cd764b1f42'
```

## Impact 
- Customer list payment methods.
- Save payment method.

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code

### CODE REVIEW

File: crates/router/src/core/payments/operations/payment_create.rs [hunk 1]
```diff
@@ -164,14 +164,34 @@ impl<F: Send + Clone, Ctx: PaymentMethodRetrieve>
...
+            if !is_payment_method_billing_address_passed {
+                payment_method_info
+                    .as_ref()
+                    .and_then(|payment_method_info| {
+                        payment_method_info
+                            .payment_method_billing_address_id
+                            .as_deref()
+                    })
+            } else {
+                None
```
Comment (sahkal) [inline crates/router/src/core/payments/operations/payment_create.rs:184]: Can we not use reverse logic

File: crates/router/src/core/payments/operations/payment_create.rs [hunk 2]
```diff
@@ -164,14 +164,34 @@ impl<F: Send + Clone, Ctx: PaymentMethodRetrieve>
         )
         .await?;
 
+        // If request billing address is not passed, use saved billing address id
+        let is_payment_method_billing_address_passed = request
+            .payment_method_data
+            .as_ref()
+            .map(|payment_method_data| payment_method_data.billing.is_some())
+            .unwrap_or(false);
```
Comment (hrithikesh026) [inline crates/router/src/core/payments/operations/payment_create.rs:172]:
```suggestion
            is_some_and(|payment_method_data| payment_method_data.billing.is_some());
```

File: crates/router/src/core/payments/operations/payment_update.rs [hunk 3]
```diff
@@ -216,13 +216,36 @@ impl<F: Send + Clone, Ctx: PaymentMethodRetrieve>
...
+                    .as_ref()
+                    .and_then(|payment_method_info| {
+                        payment_method_info
+                            .payment_method_billing_address_id
+                            .as_deref()
+                    })
+            } else {
+                None
+            };
+
```
Comment (sahkal) [inline crates/router/src/core/payments/operations/payment_update.rs:238]: +1

File: crates/router/src/core/payouts/helpers.rs [hunk 4]
```diff
@@ -493,6 +493,7 @@ pub async fn save_payout_data_to_locker(
             None,
             None,
             merchant_account.storage_scheme,
+            None,
```
Comment (sahkal) [inline crates/router/src/core/payouts/helpers.rs:496]: can you check this! why are we not doing this for payouts can we check this with @kashif-m and make the relevant change to get the billing address in payment method data object

**Review by sahkal — CHANGES_REQUESTED** [review]

**Review by hrithikesh026 — APPROVED** [review]

**Review by sai-harsha-vardhan — DISMISSED** [review]

File: crates/router/src/core/payouts/helpers.rs (hunk 4 above)
Comment (Narayanbhat166) [reply crates/router/src/core/payouts/helpers.rs:496]: This will be taken up in the next quarter, will discuss with the concerned team

**Review by Aprabhat19 — DISMISSED** [review]

**Review by sahkal — DISMISSED** [review]

**Review by sai-harsha-vardhan — APPROVED** [review]

**Review by hrithikesh026 — APPROVED** [review]

**Review by sahkal — DISMISSED** [review]

**Review by hrithikesh026 — APPROVED** [review]

**Review by sahkal — APPROVED** [review]

**Pull request merged**